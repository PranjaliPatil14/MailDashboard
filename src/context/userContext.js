import React, { useEffect, useState } from "react";
import * as PropTypes from "prop-types";
import useLocalStorage, { USER } from "../utilities/useLocalStorage";

const initialUserContextValue = {
  users: [],
  loggedInUser: {
    id: 0,
    name: "",
    role: "",
    profileImage: "",
  },
  setLoggedInUser: () => {},
  setUsers: () => {},
};

export const UserContext = React.createContext(initialUserContextValue);

const UserContextProvider = ({ children }) => {
  const [users, setUsers] = useState([]);
  const [loggedInUser, setLoggedInUser] = useState({});
  const { value } = useLocalStorage(USER);

  useEffect(() => {
    if (!users.length) {
      setUsers(value);
    }
  }, [users.length, value]);

  return (
    <UserContext.Provider
      value={{ users, setUsers, loggedInUser, setLoggedInUser }}
    >
      {children}
    </UserContext.Provider>
  );
};

UserContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default UserContextProvider;
