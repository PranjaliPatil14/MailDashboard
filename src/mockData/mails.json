[
  {"id": 1, "from":1,"to":[3],"cc":[2],"body":"Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla eget rhoncus metus.","attachment":false,"time":"Sat Jan 02 2021 18:02:02 GMT+0530 (India Standard Time)","tag":"Development", "read": false, "important": false},
  {"id": 2, "from":2,"to":[1],"cc":[3],"body":"Quisque non lectus tincidunt, consequat sapien eu, rutrum nisl.","attachment":false,"time":"Tue Jan 05 2021 07:44:34 GMT+0530 (India Standard Time)","tag":"Jobs", "read": false, "important": false},
  {"id": 3, "from":1,"to":[3],"cc":[],"body":"Phasellus ullamcorper nisl a velit vulputate, in euismod felis sollicitudin.","attachment":false,"time":"Sat Jan 02 2021 18:02:02 GMT+0530 (India Standard Time)","tag":"", "read": false, "important": false},
  {"id": 4, "from":2,"to":[1],"cc":[],"body":"Integer lacinia, ex vitae interdum lobortis, felis nisi pretium tortor, et tincidunt dolor elit sed magna.","attachment":true,"time":"Sat Jan 02 2021 18:02:02 GMT+0530 (India Standard Time)","tag":"client", "read": true, "important": true},
  {"id": 5, "from":3,"to":[2],"cc":[],"body":"Integer lacinia, ex vitae interdum lobortis, felis nisi pretium tortor, et tincidunt dolor elit sed magna.","attachment":true,"time":"Sat Jan 02 2021 18:02:02 GMT+0530 (India Standard Time)","tag":"", "read": true, "important": true},
  {"id": 6, "from":2,"to":[3],"cc":[],"body":"Integer lacinia, ex vitae interdum lobortis, felis nisi pretium tortor, et tincidunt dolor elit sed magna.","attachment":true,"time":"Sat Jan 02 2021 18:02:02 GMT+0530 (India Standard Time)","tag":"", "read": true, "important": true},
  {"id": 7, "from":3,"to":[1],"cc":[2],"body":"Integer lacinia, ex vitae interdum lobortis, felis nisi pretium tortor, et tincidunt dolor elit sed magna.","attachment":true,"time":"Sat Jan 02 2021 18:02:02 GMT+0530 (India Standard Time)","tag":"", "read": true, "important": true}
]
