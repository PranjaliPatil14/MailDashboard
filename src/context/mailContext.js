import React, { useContext, useEffect, useReducer } from "react";
import * as PropTypes from "prop-types";
import useLocalStorage, { MAIL } from "../utilities/useLocalStorage";
import { UserContext } from "./userContext";
import {
  mailReducer,
  SET_LOGGED_IN_USER_MAIL,
  SET_VALUES,
} from "./mailReducer";

const initialMailContextValue = {
  mails: [],
  refresh: true,
  loggedInUserMails: [],
  dispatch: () => {},
};

export const MailContext = React.createContext(initialMailContextValue);

const MailContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(mailReducer, initialMailContextValue);
  const { value } = useLocalStorage(MAIL);
  const { loggedInUser } = useContext(UserContext);

  useEffect(() => {
    if (state.refresh) {
      dispatch({ type: SET_VALUES, data: value });
    }
  }, [value, state.refresh]);

  useEffect(() => {
    dispatch({ type: SET_LOGGED_IN_USER_MAIL, data: loggedInUser.id });
  }, [loggedInUser, value, state.mails]);

  useEffect(() => {
    localStorage.setItem(MAIL, JSON.stringify(state.mails));
  }, [state.mails]);

  return (
    <MailContext.Provider value={{ ...state, dispatch }}>
      {children}
    </MailContext.Provider>
  );
};

MailContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default MailContextProvider;
