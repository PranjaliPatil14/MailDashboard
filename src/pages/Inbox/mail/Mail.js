import React, { useContext } from "react";
import * as PropTypes from "prop-types";
import { Paperclip } from "react-feather";
import { UserContext } from "../../../context/userContext";
import "./mail.css";

const Mail = ({ tag, body, time, attachment }) => {
  const { loggedInUser } = useContext(UserContext);

  const formatDateTime = () => {
    const monthNames = [
      "Jan",
      "Feb",
      "Mar",
      "April",
      "May",
      "Jun",
      "July",
      "Aug",
      "Sep",
      "Oct",
      "Nov",
      "Dec",
    ];
    const date = new Date(time);
    if (date.getDate() === new Date().getDate()) {
      return date.toLocaleString("en-US", {
        hour: "numeric",
        minute: "numeric",
        hour12: true,
      });
    }
    return `${monthNames[date.getMonth()]} ${date.getDate()}`;
  };

  return (
    <div className="mail">
      <input type="checkbox" />
      <p className="mail--from">{loggedInUser.name}</p>
      {tag ? (
        <div className="mail--tags">
          <span>{tag}</span>
        </div>
      ) : (
        <div />
      )}
      <p className="mail--body"> {body}</p>
      <span className="mail--attachment">
        {attachment ? <Paperclip size={12} /> : <></>}
      </span>
      <span className="mail-deliveryTime">{formatDateTime()}</span>
    </div>
  );
};

Mail.propTypes = {
  tag: PropTypes.string,
  body: PropTypes.string.isRequired,
  time: PropTypes.string.isRequired,
  attachment: PropTypes.bool,
};

Mail.defaultProps = {
  tag: "",
  attachment: false,
};

export default Mail;
