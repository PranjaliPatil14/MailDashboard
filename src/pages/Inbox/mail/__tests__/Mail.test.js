import React from "react";
import { render, fireEvent } from "@testing-library/react";
import mockUser from "../../../../mockData/users.json";
import mails from "../../../../mockData/mails.json";
import Mail from "../Mail";
import { UserContext } from "../../../../context/userContext";

describe("MailHeader", () => {
  const setSelectedMails = jest.fn().mockImplementation((c) => c([]));
  const users = mockUser;
  let component;
  const renderMail = (mail) => {
    component = render(
      <UserContext.Provider value={{ users }}>
        <Mail setSelectedMails={setSelectedMails} mail={mail} />
      </UserContext.Provider>
    );
  };

  test("should render mail", () => {
    renderMail(mails[0]);
    expect(component.asFragment()).toMatchSnapshot();
  });

  test("should render important mail", () => {
    renderMail(mails[4]);
    expect(component.asFragment()).toMatchSnapshot();
  });

  test("should render mail with current date as email time", () => {
    const date = new Date();
    renderMail({ ...mails[4], time: date.toString() });
    expect(
      component.getByText(
        date.toLocaleString("en-US", {
          hour: "numeric",
          minute: "numeric",
          hour12: true,
        })
      )
    ).toBeInTheDocument();
  });

  test("should call setSelectedMails on checkbox check and uncheck", () => {
    renderMail(mails[4]);
    fireEvent.click(component.getByRole("checkbox"));

    expect(setSelectedMails).toHaveBeenCalled();

    fireEvent.click(component.getByRole("checkbox"));

    expect(setSelectedMails).toHaveBeenCalled();
  });
});
