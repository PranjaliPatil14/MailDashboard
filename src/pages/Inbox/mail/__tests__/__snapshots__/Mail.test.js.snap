// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MailHeader should render important mail 1`] = `
<DocumentFragment>
  <div
    class="mail read"
  >
    <input
      type="checkbox"
    />
    <div
      class="mail--important"
    >
      <svg
        fill="none"
        height="16"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polygon
          points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
        />
      </svg>
    </div>
    <p
      class="mail--from"
    >
      mat
    </p>
    <div />
    <p
      class="mail--body"
    >
       Integer lacinia, ex vitae interdum lobortis, felis nisi pretium tortor, et tincidunt dolor elit sed magna.
    </p>
    <span
      class="mail--attachment"
    >
      <svg
        fill="none"
        height="12"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="12"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M21.44 11.05l-9.19 9.19a6 6 0 0 1-8.49-8.49l9.19-9.19a4 4 0 0 1 5.66 5.66l-9.2 9.19a2 2 0 0 1-2.83-2.83l8.49-8.48"
        />
      </svg>
    </span>
    <span
      class="mail-deliveryTime"
    >
      Jan 2
    </span>
  </div>
</DocumentFragment>
`;

exports[`MailHeader should render mail 1`] = `
<DocumentFragment>
  <div
    class="mail"
  >
    <input
      type="checkbox"
    />
    <div
      class="mail--important"
    />
    <p
      class="mail--from"
    >
      abc
    </p>
    <div
      class="mail--tags"
    >
      <span>
        Development
      </span>
    </div>
    <p
      class="mail--body"
    >
       Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla eget rhoncus metus.
    </p>
    <span
      class="mail--attachment"
    />
    <span
      class="mail-deliveryTime"
    >
      Jan 2
    </span>
  </div>
</DocumentFragment>
`;
